src/derivative_utils.jl

```diff
@@ -351,14 +364,16 @@ end
   return W
 end
 
-function calc_W!(W, integrator, cache, dtgamma, repeat_step, W_transform=false)
+function calc_W!(W, integrator, nlsolver::Union{Nothing,AbstractNLSolver}, cache, dtgamma, repeat_step, W_transform=false)
   @unpack t,dt,uprev,u,f,p = integrator
-  @unpack J = cache
+  lcache = nlsolver === nothing ? cache : nlsolver.cache
+  @unpack J = lcache
   alg = unwrap_alg(integrator, true)
   mass_matrix = integrator.f.mass_matrix
   is_compos = integrator.alg isa CompositeAlgorithm
   isnewton = alg isa NewtonAlgorithm
 
+  # handle Wfact
   if W_transform && DiffEqBase.has_Wfact_t(f)
```

Comment:
Do these even if a new `W` might not be needed? I guess the assumption is that it's cheap enough so that if it exists, then you might as well skip jacobian reuse, but that's not true with DiffEqGPU which wants to use `Wfact` but also have Jacobian reuse in its large factorizations.